### PR TITLE
MD Team Policies Implementation for cuda

### PR DESCRIPTION
**Pull request by ldh4**
Implementation for MD TeamPolicy for Cuda backend.
See #4812 

@nliber

### CODE REVIEW

File: core/src/Cuda/Kokkos_Cuda_Team.hpp [hunk 1]
```diff
@@ -417,6 +417,98 @@ struct ThreadVectorRangeBoundariesStruct<iType, CudaTeamMember> {
...
+  template <typename... Ns>
+  KOKKOS_INLINE_FUNCTION constexpr explicit MDTeamVectorRangeBoundariesStruct(
+      team_member_type const& tm, Ns&&... ns)
+      : team_member(tm), taskDims{static_cast<iType>(ns)...} {
+    static_assert(sizeof...(ns) == Rank, "Number of ns must equal Rank");
+  }
+
+  team_member_type const& team_member;
+  iType const taskDims[Rank];
+};
```
Comment (masterleinad) [inline core/src/Cuda/Kokkos_Cuda_Team.hpp:510]: and here
```suggestion
```

**Comment by masterleinad** [comment]
You will probably also need something like
```diff
diff --git a/core/unit_test/TestMDTeamParallelism.hpp b/core/unit_test/TestMDTeamParallelism.hpp
index 378474998..b87ce7ccb 100644
--- a/core/unit_test/TestMDTeamParallelism.hpp
+++ b/core/unit_test/TestMDTeamParallelism.hpp
@@ -108,9 +108,10 @@ struct TestMDTeamParallelFor {
 
           auto teamRange = Kokkos::MDTeamThreadRange<Direction>(team, n0, n1);
 
-          Kokkos::parallel_for(teamRange, [=](int i, int j) {
-            v(leagueRank, i, j) += fillFlattenedIndex(leagueRank, i, j);
-          });
+          Kokkos::parallel_for(
+              teamRange, KOKKOS_LAMBDA(int i, int j) {
+                v(leagueRank, i, j) += fillFlattenedIndex(leagueRank, i, j);
+              });
         });
 
     HostViewType h_view = Kokkos::create_mirror_view_and_copy(
@@ -153,9 +154,11 @@ struct TestMDTeamParallelFor {
           auto teamRange =
               Kokkos::MDTeamThreadRange<Direction>(team, n0, n1, n2);
 
-          Kokkos::parallel_for(teamRange, [=](int i, int j, int k) {
-            v(leagueRank, i, j, k) += fillFlattenedIndex(leagueRank, i, j, k);
-          });
+          Kokkos::parallel_for(
+              teamRange, KOKKOS_LAMBDA(int i, int j, int k) {
+                v(leagueRank, i, j, k) +=
+                    fillFlattenedIndex(leagueRank, i, j, k);
+              });
         });
 
     HostViewType h_view = Kokkos::create_mirror_view_and_copy(
@@ -201,10 +204,11 @@ struct TestMDTeamParallelFor {
           auto teamRange =
               Kokkos::MDTeamThreadRange<Direction>(team, n0, n1, n2, n3);
 
-          Kokkos::parallel_for(teamRange, [=](int i, int j, int k, int l) {
-            v(leagueRank, i, j, k, l) +=
-                fillFlattenedIndex(leagueRank, i, j, k, l);
-          });
+          Kokkos::parallel_for(
+              teamRange, KOKKOS_LAMBDA(int i, int j, int k, int l) {
+                v(leagueRank, i, j, k, l) +=
+                    fillFlattenedIndex(leagueRank, i, j, k, l);
+              });
         });
 
     HostViewType h_view = Kokkos::create_mirror_view_and_copy(
@@ -255,7 +259,7 @@ struct TestMDTeamParallelFor {
               Kokkos::MDTeamThreadRange<Direction>(team, n0, n1, n2, n3, n4);
 
           Kokkos::parallel_for(
-              teamRange, [=](int i, int j, int k, int l, int m) {
+              teamRange, KOKKOS_LAMBDA(int i, int j, int k, int l, int m) {
                 v(leagueRank, i, j, k, l, m) +=
                     fillFlattenedIndex(leagueRank, i, j, k, l, m);
               });
@@ -312,7 +316,8 @@ struct TestMDTeamParallelFor {
                                                                 n2, n3, n4, n5);
 
           Kokkos::parallel_for(
-              teamRange, [=](int i, int j, int k, int l, int m, int n) {
+              teamRange,
+              KOKKOS_LAMBDA(int i, int j, int k, int l, int m, int n) {
                 v(leagueRank, i, j, k, l, m, n) +=
                     fillFlattenedIndex(leagueRank, i, j, k, l, m, n);
               });
@@ -373,7 +378,8 @@ struct TestMDTeamParallelFor {
               team, n0, n1, n2, n3, n4, n5, n6);
 
           Kokkos::parallel_for(
-              teamRange, [=](int i, int j, int k, int l, int m, int n, int o) {
+              teamRange,
+              KOKKOS_LAMBDA(int i, int j, int k, int l, int m, int n, int o) {
                 v(leagueRank, i, j, k, l, m, n, o) +=
                     fillFlattenedIndex(leagueRank, i, j, k, l, m, n, o);
               });
@@ -409,11 +415,14 @@ struct TestMDTeamParallelFor {
               Kokkos::MDThreadVectorRange<OuterDirection, InnerDirection>(
                   team, n1, n2);
 
-          Kokkos::parallel_for(teamThreadRange, [=](int i) {
-            Kokkos::parallel_for(teamRange, [=](int j, int k) {
-              v(leagueRank, i, j, k) += fillFlattenedIndex(leagueRank, i, j, k);
-            });
-          });
+          Kokkos::parallel_for(
+              teamThreadRange, KOKKOS_LAMBDA(int i) {
+                Kokkos::parallel_for(
+                    teamRange, KOKKOS_LAMBDA(int j, int k) {
+                      v(leagueRank, i, j, k) +=
+                          fillFlattenedIndex(leagueRank, i, j, k);
+                    });
+              });
         });
 
     HostViewType h_view = Kokkos::create_mirror_view_and_copy(
@@ -447,12 +456,14 @@ struct TestMDTeamParallelFor {
               Kokkos::MDThreadVectorRange<OuterDirection, InnerDirection>(
                   team, n1, n2, n3);
 
-          Kokkos::parallel_for(teamThreadRange, [=](int i) {
-            Kokkos::parallel_for(teamRange, [=](int j, int k, int l) {
-              v(leagueRank, i, j, k, l) +=
-                  fillFlattenedIndex(leagueRank, i, j, k, l);
-            });
-          });
+          Kokkos::parallel_for(
+              teamThreadRange, KOKKOS_LAMBDA(int i) {
+                Kokkos::parallel_for(
+                    teamRange, KOKKOS_LAMBDA(int j, int k, int l) {
+                      v(leagueRank, i, j, k, l) +=
+                          fillFlattenedIndex(leagueRank, i, j, k, l);
+                    });
+              });
         });
 
     HostViewType h_view = Kokkos::create_mirror_view_and_copy(
@@ -487,12 +498,14 @@ struct TestMDTeamParallelFor {
               Kokkos::MDThreadVectorRange<OuterDirection, InnerDirection>(
                   team, n1, n2, n3, n4);
 
-          Kokkos::parallel_for(teamThreadRange, [=](int i) {
-            Kokkos::parallel_for(teamRange, [=](int j, int k, int l, int m) {
-              v(leagueRank, i, j, k, l, m) +=
-                  fillFlattenedIndex(leagueRank, i, j, k, l, m);
-            });
-          });
+          Kokkos::parallel_for(
+              teamThreadRange, KOKKOS_LAMBDA(int i) {
+                Kokkos::parallel_for(
+                    teamRange, KOKKOS_LAMBDA(int j, int k, int l, int m) {
+                      v(leagueRank, i, j, k, l, m) +=
+                          fillFlattenedIndex(leagueRank, i, j, k, l, m);
+                    });
+              });
         });
 
     HostViewType h_view = Kokkos::create_mirror_view_and_copy(
@@ -528,13 +541,15 @@ struct TestMDTeamParallelFor {
               Kokkos::MDThreadVectorRange<OuterDirection, InnerDirection>(
                   team, n1, n2, n3, n4, n5);
 
-          Kokkos::parallel_for(teamThreadRange, [=](int i) {
-            Kokkos::parallel_for(
-                teamRange, [=](int j, int k, int l, int m, int n) {
-                  v(leagueRank, i, j, k, l, m, n) +=
-                      fillFlattenedIndex(leagueRank, i, j, k, l, m, n);
-                });
-          });
+          Kokkos::parallel_for(
+              teamThreadRange, KOKKOS_LAMBDA(int i) {
+                Kokkos::parallel_for(
+                    teamRange,
+                    KOKKOS_LAMBDA(int j, int k, int l, int m, int n) {
+                      v(leagueRank, i, j, k, l, m, n) +=
+                          fillFlattenedIndex(leagueRank, i, j, k, l, m, n);
+                    });
+              });
         });
 
     HostViewType h_view = Kokkos::create_mirror_view_and_copy(
@@ -572,13 +587,15 @@ struct TestMDTeamParallelFor {
               Kokkos::MDThreadVectorRange<OuterDirection, InnerDirection>(
                   team, n1, n2, n3, n4, n5, n6);
 
-          Kokkos::parallel_for(teamThreadRange, [=](int i) {
-            Kokkos::parallel_for(
-                teamRange, [=](int j, int k, int l, int m, int n, int o) {
-                  v(leagueRank, i, j, k, l, m, n, o) +=
-                      fillFlattenedIndex(leagueRank, i, j, k, l, m, n, o);
-                });
-          });
+          Kokkos::parallel_for(
+              teamThreadRange, KOKKOS_LAMBDA(int i) {
+                Kokkos::parallel_for(
+                    teamRange,
+                    KOKKOS_LAMBDA(int j, int k, int l, int m, int n, int o) {
+                      v(leagueRank, i, j, k, l, m, n, o) +=
+                          fillFlattenedIndex(leagueRank, i, j, k, l, m, n, o);
+                    });
+              });
         });
 
     HostViewType h_view = Kokkos::create_mirror_view_and_copy(
@@ -609,9 +626,10 @@ struct TestMDTeamParallelFor {
               Kokkos::MDTeamVectorRange<OuterDirection, InnerDirection>(team,
                                                                         n0, n1);
 
-          Kokkos::parallel_for(teamRange, [=](int i, int j) {
-            v(leagueRank, i, j) += fillFlattenedIndex(leagueRank, i, j);
-          });
+          Kokkos::parallel_for(
+              teamRange, KOKKOS_LAMBDA(int i, int j) {
+                v(leagueRank, i, j) += fillFlattenedIndex(leagueRank, i, j);
+              });
         });
 
     HostViewType h_view = Kokkos::create_mirror_view_and_copy(
@@ -643,9 +661,11 @@ struct TestMDTeamParallelFor {
               Kokkos::MDTeamVectorRange<OuterDirection, InnerDirection>(
                   team, n0, n1, n2);
 
-          Kokkos::parallel_for(teamRange, [=](int i, int j, int k) {
-            v(leagueRank, i, j, k) += fillFlattenedIndex(leagueRank, i, j, k);
-          });
+          Kokkos::parallel_for(
+              teamRange, KOKKOS_LAMBDA(int i, int j, int k) {
+                v(leagueRank, i, j, k) +=
+                    fillFlattenedIndex(leagueRank, i, j, k);
+              });
         });
 
     HostViewType h_view = Kokkos::create_mirror_view_and_copy(
@@ -678,10 +698,11 @@ struct TestMDTeamParallelFor {
               Kokkos::MDTeamVectorRange<OuterDirection, InnerDirection>(
                   team, n0, n1, n2, n3);
 
-          Kokkos::parallel_for(teamRange, [=](int i, int j, int k, int l) {
-            v(leagueRank, i, j, k, l) +=
-                fillFlattenedIndex(leagueRank, i, j, k, l);
-          });
+          Kokkos::parallel_for(
+              teamRange, KOKKOS_LAMBDA(int i, int j, int k, int l) {
+                v(leagueRank, i, j, k, l) +=
+                    fillFlattenedIndex(leagueRank, i, j, k, l);
+              });
         });
 
     HostViewType h_view = Kokkos::create_mirror_view_and_copy(
@@ -716,7 +737,7 @@ struct TestMDTeamParallelFor {
                   team, n0, n1, n2, n3, n4);
 
           Kokkos::parallel_for(
-              teamRange, [=](int i, int j, int k, int l, int m) {
+              teamRange, KOKKOS_LAMBDA(int i, int j, int k, int l, int m) {
                 v(leagueRank, i, j, k, l, m) +=
                     fillFlattenedIndex(leagueRank, i, j, k, l, m);
               });
@@ -755,7 +776,8 @@ struct TestMDTeamParallelFor {
                   team, n0, n1, n2, n3, n4, n5);
 
           Kokkos::parallel_for(
-              teamRange, [=](int i, int j, int k, int l, int m, int n) {
+              teamRange,
+              KOKKOS_LAMBDA(int i, int j, int k, int l, int m, int n) {
                 v(leagueRank, i, j, k, l, m, n) +=
                     fillFlattenedIndex(leagueRank, i, j, k, l, m, n);
               });
@@ -796,7 +818,8 @@ struct TestMDTeamParallelFor {
                   team, n0, n1, n2, n3, n4, n5, n6);
 
           Kokkos::parallel_for(
-              teamRange, [=](int i, int j, int k, int l, int m, int n, int o) {
+              teamRange,
+              KOKKOS_LAMBDA(int i, int j, int k, int l, int m, int n, int o) {
                 v(leagueRank, i, j, k, l, m, n, o) +=
                     fillFlattenedIndex(leagueRank, i, j, k, l, m, n, o);
               });
@@ -826,9 +849,10 @@ struct TestMDTeamParallelFor {
           auto teamRange =
               Kokkos::MDTeamThreadRange<Direction>(team, n0, n1, n2);
 
-          Kokkos::parallel_for(teamRange, [=](int i, int j, int k) {
-            v(i, j, k) += fillFlattenedIndex(i, j, k);
-          });
+          Kokkos::parallel_for(
+              teamRange, KOKKOS_LAMBDA(int i, int j, int k) {
+                v(i, j, k) += fillFlattenedIndex(i, j, k);
+              });
         });
 
     HostViewType h_view = Kokkos::create_mirror_view_and_copy(
@@ -858,9 +882,10 @@ struct TestMDTeamParallelFor {
               Kokkos::MDTeamVectorRange<OuterDirection, InnerDirection>(
                   team, n0, n1, n2, n3);
 
-          Kokkos::parallel_for(teamRange, [=](int i, int j, int k, int l) {
-            v(i, j, k, l) += fillFlattenedIndex(i, j, k, l);
-          });
+          Kokkos::parallel_for(
+              teamRange, KOKKOS_LAMBDA(int i, int j, int k, int l) {
+                v(i, j, k, l) += fillFlattenedIndex(i, j, k, l);
+              });
         });
 
     HostViewType h_view = Kokkos::create_mirror_view_and_copy(
@@ -931,7 +956,7 @@ struct TestMDTeamParallelReduce {
 
           Kokkos::parallel_reduce(
               Kokkos::MDTeamThreadRange<Direction>(team, n0, n1),
-              [=](const int& i, const int& j, int& threadSum) {
+              KOKKOS_LAMBDA(const int& i, const int& j, int& threadSum) {
                 threadSum += v(leagueRank, i, j);
               },
               teamSum);
@@ -976,7 +1001,8 @@ struct TestMDTeamParallelReduce {
 
           Kokkos::parallel_reduce(
               Kokkos::MDTeamThreadRange<Direction>(team, n0, n1, n2),
-              [=](const int& i, const int& j, const int& k, int& threadSum) {
+              KOKKOS_LAMBDA(const int& i, const int& j, const int& k,
+                            int& threadSum) {
                 threadSum += v(leagueRank, i, j, k);
               },
               teamSum);
@@ -1023,8 +1049,10 @@ struct TestMDTeamParallelReduce {
 
           Kokkos::parallel_reduce(
               Kokkos::MDTeamThreadRange<Direction>(team, n0, n1, n2, n3),
-              [=](const int& i, const int& j, const int& k, const int& l,
-                  int& threadSum) { threadSum += v(leagueRank, i, j, k, l); },
+              KOKKOS_LAMBDA(const int& i, const int& j, const int& k,
+                            const int& l, int& threadSum) {
+                threadSum += v(leagueRank, i, j, k, l);
+              },
               teamSum);
 
           Kokkos::single(Kokkos::PerTeam(team),
@@ -1070,8 +1098,8 @@ struct TestMDTeamParallelReduce {
 
           Kokkos::parallel_reduce(
               Kokkos::MDTeamThreadRange<Direction>(team, n0, n1, n2, n3, n4),
-              [=](const int& i, const int& j, const int& k, const int& l,
-                  const int& m, int& threadSum) {
+              KOKKOS_LAMBDA(const int& i, const int& j, const int& k,
+                            const int& l, const int& m, int& threadSum) {
                 threadSum += v(leagueRank, i, j, k, l, m);
               },
               teamSum);
@@ -1127,8 +1155,9 @@ struct TestMDTeamParallelReduce {
           Kokkos::parallel_reduce(
               Kokkos::MDTeamThreadRange<Direction>(team, n0, n1, n2, n3, n4,
                                                    n5),
-              [=](const int& i, const int& j, const int& k, const int& l,
-                  const int& m, const int& n, int& threadSum) {
+              KOKKOS_LAMBDA(const int& i, const int& j, const int& k,
+                            const int& l, const int& m, const int& n,
+                            int& threadSum) {
                 threadSum += v(leagueRank, i, j, k, l, m, n);
               },
               teamSum);
@@ -1186,8 +1215,9 @@ struct TestMDTeamParallelReduce {
           Kokkos::parallel_reduce(
               Kokkos::MDTeamThreadRange<Direction>(team, n0, n1, n2, n3, n4, n5,
                                                    n6),
-              [=](const int& i, const int& j, const int& k, const int& l,
-                  const int& m, const int& n, const int& o, int& threadSum) {
+              KOKKOS_LAMBDA(const int& i, const int& j, const int& k,
+                            const int& l, const int& m, const int& n,
+                            const int& o, int& threadSum) {
                 threadSum += v(leagueRank, i, j, k, l, m, n, o);
               },
               teamSum);
@@ -1240,7 +1270,7 @@ struct TestMDTeamParallelReduce {
             int threadSum = 0;
             Kokkos::parallel_reduce(
                 threadVectorRange,
-                [=](const int& j, const int& k, int& vectorSum) {
+                KOKKOS_LAMBDA(const int& j, const int& k, int& vectorSum) {
                   vectorSum += v(leagueRank, i, j, k);
                 },
                 threadSum);
@@ -1297,7 +1327,8 @@ struct TestMDTeamParallelReduce {
             int threadSum = 0;
             Kokkos::parallel_reduce(
                 threadVectorRange,
-                [=](const int& j, const int& k, const int& l, int& vectorSum) {
+                KOKKOS_LAMBDA(const int& j, const int& k, const int& l,
+                              int& vectorSum) {
                   vectorSum += v(leagueRank, i, j, k, l);
                 },
                 threadSum);
@@ -1355,8 +1386,8 @@ struct TestMDTeamParallelReduce {
             int threadSum = 0;
             Kokkos::parallel_reduce(
                 threadVectorRange,
-                [=](const int& j, const int& k, const int& l, const int& m,
-                    int& vectorSum) {
+                KOKKOS_LAMBDA(const int& j, const int& k, const int& l,
+                              const int& m, int& vectorSum) {
                   vectorSum += v(leagueRank, i, j, k, l, m);
                 },
                 threadSum);
@@ -1419,8 +1450,8 @@ struct TestMDTeamParallelReduce {
             int threadSum = 0;
             Kokkos::parallel_reduce(
                 threadVectorRange,
-                [=](const int& j, const int& k, const int& l, const int& m,
-                    const int& n, int& vectorSum) {
+                KOKKOS_LAMBDA(const int& j, const int& k, const int& l,
+                              const int& m, const int& n, int& vectorSum) {
                   vectorSum += v(leagueRank, i, j, k, l, m, n);
                 },
                 threadSum);
@@ -1487,8 +1518,9 @@ struct TestMDTeamParallelReduce {
             int threadSum = 0;
             Kokkos::parallel_reduce(
                 threadVectorRange,
-                [=](const int& j, const int& k, const int& l, const int& m,
-                    const int& n, const int& o, int& vectorSum) {
+                KOKKOS_LAMBDA(const int& j, const int& k, const int& l,
+                              const int& m, const int& n, const int& o,
+                              int& vectorSum) {
                   vectorSum += v(leagueRank, i, j, k, l, m, n, o);
                 },
                 threadSum);
@@ -1540,7 +1572,8 @@ struct TestMDTeamParallelReduce {
 
           Kokkos::parallel_reduce(
               teamVectorRange,
-              [=](const int& i, const int& j, const int& k, int& vectorSum) {
+              KOKKOS_LAMBDA(const int& i, const int& j, const int& k,
+                            int& vectorSum) {
                 vectorSum += v(leagueRank, i, j, k);
               },
               teamSum);
@@ -1592,8 +1625,10 @@ struct TestMDTeamParallelReduce {
 
           Kokkos::parallel_reduce(
               teamVectorRange,
-              [=](const int& i, const int& j, const int& k, const int& l,
-                  int& vectorSum) { vectorSum += v(leagueRank, i, j, k, l); },
+              KOKKOS_LAMBDA(const int& i, const int& j, const int& k,
+                            const int& l, int& vectorSum) {
+                vectorSum += v(leagueRank, i, j, k, l);
+              },
               teamSum);
 
           Kokkos::single(Kokkos::PerTeam(team),
@@ -1644,8 +1679,8 @@ struct TestMDTeamParallelReduce {
 
           Kokkos::parallel_reduce(
               teamVectorRange,
-              [=](const int& i, const int& j, const int& k, const int& l,
-                  const int& m, int& vectorSum) {
+              KOKKOS_LAMBDA(const int& i, const int& j, const int& k,
+                            const int& l, const int& m, int& vectorSum) {
                 vectorSum += v(leagueRank, i, j, k, l, m);
               },
               teamSum);
@@ -1703,8 +1738,9 @@ struct TestMDTeamParallelReduce {
 
           Kokkos::parallel_reduce(
               teamVectorRange,
-              [=](const int& i, const int& j, const int& k, const int& l,
-                  const int& m, const int& n, int& vectorSum) {
+              KOKKOS_LAMBDA(const int& i, const int& j, const int& k,
+                            const int& l, const int& m, const int& n,
+                            int& vectorSum) {
                 vectorSum += v(leagueRank, i, j, k, l, m, n);
               },
               teamSum);
@@ -1766,8 +1802,9 @@ struct TestMDTeamParallelReduce {
 
           Kokkos::parallel_reduce(
               teamVectorRange,
-              [=](const int& i, const int& j, const int& k, const int& l,
-                  const int& m, const int& n, const int& o, int& vectorSum) {
+              KOKKOS_LAMBDA(const int& i, const int& j, const int& k,
+                            const int& l, const int& m, const int& n,
+                            const int& o, int& vectorSum) {
                 vectorSum += v(leagueRank, i, j, k, l, m, n, o);
               },
               teamSum);
```
for the test to work with `Cuda`, right?

**Review by masterleinad — COMMENTED** [review]
Fine with me when CI passes.

File: core/src/Cuda/Kokkos_Cuda_Team.hpp [hunk 2]
```diff
@@ -417,6 +417,98 @@ struct ThreadVectorRangeBoundariesStruct<iType, CudaTeamMember> {
...
+  KOKKOS_INLINE_FUNCTION
+  MDTeamThreadRangeBoundariesStruct(team_member_type const& member,
+                                    const iType (&array)[Rank])
+      : thread(member) {
+    std::copy(&array[0], &array[Rank], &threadDims[0]);
+  }
+
+  team_member_type const& thread;
+  iType threadDims[Rank];
+};
```
[the remark this answers is not in the document]
Comment (ldh4) [reply core/src/Cuda/Kokkos_Cuda_Team.hpp]: Resolved in [5114b38](https://github.com/kokkos/kokkos/commit/5114b38d817e6b399f4ba686af0f16cdf1e03039)

File: core/src/Cuda/Kokkos_Cuda_Team.hpp [hunk 3]
```diff
@@ -417,6 +417,98 @@ struct ThreadVectorRangeBoundariesStruct<iType, CudaTeamMember> {
...
+  template <typename... Ns>
+  KOKKOS_INLINE_FUNCTION constexpr explicit MDThreadVectorRangeBoundariesStruct(
+      team_member_type const& tm, Ns&&... ns)
+      : team_member(tm), taskDims{static_cast<iType>(ns)...} {
+    static_assert(sizeof...(ns) == Rank, "Number of ns must equal Rank");
+  }
+
+  team_member_type const& team_member;
+  iType const taskDims[Rank];
+};
```
[the remark this answers is not in the document]
Comment (ldh4) [reply core/src/Cuda/Kokkos_Cuda_Team.hpp]: Resolved in [5114b38](https://github.com/kokkos/kokkos/commit/5114b38d817e6b399f4ba686af0f16cdf1e03039)

File: core/src/Cuda/Kokkos_Cuda_Team.hpp (hunk 1 above)
Comment (ldh4) [reply core/src/Cuda/Kokkos_Cuda_Team.hpp]: Resolved in [5114b38](https://github.com/kokkos/kokkos/commit/5114b38d817e6b399f4ba686af0f16cdf1e03039)

**Comment by ldh4** [comment]
@masterleinad For those suggested changes, I thought extended lambda cannot be nested in another extended lambda.

**Comment by masterleinad** [comment]
> @masterleinad For those suggested changes, I thought extended lambda cannot be nested in another extended lambda.

Oh, d'oh! You're right of course. Then, I'm not quite sure what the `Cuda` C++17 build is complaining about.

**Comment by crtrott** [comment]
I think there was a general misunderstanding about what left vs right means. Its like in MDRangePolicy about which index gets iterated over fastest. I.e. whether the left most or right most is the one which is the fastest index within a thread. The basic driving idea is that a LayoutLeft view iterated over with a LayoutLeft MDRange should give optimal access pattern. We can talk Monday.

**Pull request closed without merging**